web/app/js/components/ResourceList.jsx

```diff
@@ -6,114 +6,48 @@ import MetricsTable from './MetricsTable.jsx';
-    super(props);
-    this.api = this.props.api;
-    this.handleApiError = this.handleApiError.bind(this);
-    this.loadFromServer = this.loadFromServer.bind(this);
-
-    this.state = this.getInitialState(this.props);
-  }
-
-  getInitialState() {
-    return {
-      pollingInterval: 2000, // TODO: poll based on metricsWindow size
-      metrics: [],
-      pendingRequests: false,
-      loaded: false,
-      error: ''
-    };
-  }
-
-  componentDidMount() {
-    this.startServerPolling(this.props.resource);
-  }
-
-  componentWillReceiveProps(newProps) {
-    // React won't unmount this component when switching resource pages so we need to clear state
-    this.stopServerPolling();
-    this.setState(this.getInitialState(newProps));
-    this.startServerPolling(newProps.resource);
-  }
-
-  componentWillUnmount() {
-    this.stopServerPolling();
-  }
-
-  startServerPolling(resource) {
-    this.loadFromServer(resource);
-    this.timerId = window.setInterval(this.loadFromServer, this.state.pollingInterval, resource);
-  }
-
-  stopServerPolling() {
-    window.clearInterval(this.timerId);
-    this.api.cancelCurrentRequests();
-  }
-
-  loadFromServer(resource) {
-    if (this.state.pendingRequests) {
-      return; // don't make more requests if the ones we sent haven't completed
-    }
-    this.setState({ pendingRequests: true });
-
-    this.api.setCurrentRequests([
-      this.api.fetchMetrics(this.api.urlsForResource(resource))
-    ]);
-
-    Promise.all(this.api.getCurrentPromises())
-      .then(([rollup]) => {
-        let processedMetrics = processSingleResourceRollup(rollup, this.props.controllerNamespace);
-
-        this.setState({
-          metrics: processedMetrics,
-          loaded: true,
-          pendingRequests: false,
-          error: ''
-        });
-      })
-      .catch(this.handleApiError);
-  }
-
-  handleApiError(e) {
-    if (e.isCanceled) {
-      return;
-    }
-
-    this.setState({
-      pendingRequests: false,
-      error: `Error getting data from server: ${e.message}`
-    });
-  }
-
-  renderEmptyMessage() {
+export class ResourceList extends React.Component {
+  renderEmptyMessage = () => {
     let shortResource = this.props.resource === "replication_controller" ?
       "RC" : this.props.resource;
     return (<CallToAction
       resource={shortResource}
-      numResources={_.size(this.state.metrics)} />);
+      numResources={0} />);
   }
 
   render() {
-    let friendlyTitle = _.startCase(this.props.resource);
+    const {api, controllerNamespace, data, error, loading, resource} = this.props;
+
+    if (error) return  <ErrorBanner message={error} />;
+    if (loading) return <ConduitSpinner />;
+
+    const processedMetrics = processSingleResourceRollup(
+      data[0], controllerNamespace);
+
+    const friendlyTitle = _.startCase(resource);
     return (
       <div className="page-content">
-        { !this.state.error ? null : <ErrorBanner message={this.state.error} /> }
```

Comment:
The previous implementation of this component was setup so that, if an API error was encountered, we'd still display the page header, below the error banner. In this branch you're changing it so that we only display the error banner and omit the page header, by returning early on line 25. It now looks like this:

![image](https://user-images.githubusercontent.com/9226/40462008-dec62160-5ec2-11e8-9400-9b43024112af.png)

I have a slight preference for the previous implementation, so I'd vote to remove the early return and restore this line instead.

Reply:
You're totally right, this is why I hate that syntax style ... it is super hard to parse. Let me give it another go and see what you think.

Reply:
@klingerf I updated and added some tests for `ResourceList` to make sure that is the validated behavior. Mind taking a quick look and making sure that's good?